lxd/proxy_device_utils.go

```diff
@@ -75,19 +80,65 @@ func setupProxyProcInfo(c container, device map[string]string) (*proxyProcInfo,
 	return p, nil
 }
 
-func killProxyProc(devPath string) error {
-	contents, err := ioutil.ReadFile(devPath)
+func killProxyProc(pidPath string) error {
+	// Get the contents of the pid file
+	contents, err := ioutil.ReadFile(pidPath)
 	if err != nil {
 		return err
 	}
+	pidString := strings.TrimSpace(string(contents))
+
+	// Check if the process still exists
+	if !shared.PathExists(fmt.Sprintf("/proc/%s", pidString)) {
+		os.Remove(pidPath)
+		return nil
+	}
 
-	pid, _ := strconv.Atoi(string(contents))
+	// Check if it's a proxy process
+	cmdPath, err := os.Readlink(fmt.Sprintf("/proc/%s/exe", pidString))
+	if err != nil {
+		cmdPath = ""
+	}
+
+	// Deal with deleted paths
+	cmdName := filepath.Base(strings.Split(cmdPath, " ")[0])
+	if cmdName != "lxd" {
+		os.Remove(pidPath)
+		return nil
+	}
+
+	// Parse the pid
+	pidInt, _ := strconv.Atoi(pidString)
 	if err != nil {
 		return err
 	}
 
-	syscall.Kill(pid, syscall.SIGINT)
-	os.Remove(devPath)
+	err = syscall.Kill(pidInt, syscall.SIGTERM)
+	if err != nil {
```

Comment:
That looks wrong. If SIGTERM doesn't kill the process you won't get an error back.

You'd only get an error back from syscall.Kill if the PID you provided doesn't exist or you don't have the necessary rights to send the signal. If the process exists, then it Kill will always succeed, regardless of whether the process is doing something with the signal.

Signals on Unix systems are effectively just a bitmask that the process reads when notified, it's completely asynchronous and doesn't support queuing.

So the logic here should be to return any error you get from Kill.
If you don't get an error (the signal was sent properly), then enter a pretty tight loop that will check if the process has now gone away, if not wait for 500ms or so and check again, until you reach the 3s timeout at which point you just send SIGTERM and return.